sqlwhat/State.py

```diff
@@ -30,6 +31,28 @@ def __init__(self,
         if solution_ast is None: self.solution_ast = self.ast_dispatcher.parse(solution_code)
         if student_ast  is None: self.student_ast  = self.ast_dispatcher.parse(student_code)
 
+    def get_ast_path(self):
+        rev_checks = filter(lambda x: x['type'] in ['check_field', 'check_node'], reversed(self.history))
+
+        try:
+            last = next(rev_checks)
+            if last['type'] == 'check_node':
+                # final check was for a node
+                return self.ast_dispatcher.describe(last['node'],
+                                                    index = last['kwargs']['index'],
+                                                    msg = "{index}{node_name}")
+            else:
+                node = next(rev_checks)
+                if node['type'] == 'check_node':
+                    # checked for node, then for target, so can give rich description
+                    return self.ast_dispatcher.describe(node['node'],
+                                                        field = last['kwargs']['name'],
+                                                        index = last['kwargs']['index'],
+                                                        msg = "{index}{field_name} of the {node_name}")
+        except StopIteration:
```

Comment:
it's raised whenever the iterator runs out of things to yield. We use `next` over the reversed sequence to get the last entry, and then use it later to get the second to last. But if the sequence doesn't have a next value, we want to bail out.